doc/architectural/central-data-structures.md

```diff
@@ -0,0 +1,91 @@
+CBMC and the assorted CProver tools.
+
+## goto_modelt
+
+The `goto_modelt` is the main data structure that CBMC (and the other tools) use to
+represent GOTO-IR (the `GOTO` Intermediate Representation).
+
+A `goto_modelt` is effectively a product type, consisting of:
+
+* A list of GOTO functions (pseudocode: `type goto_functionst = list<goto_functiont>`)
+* A symbol table containing symbol references for the symbols contained in the
+  GOTO functions (pseudocode: `type symbol_tablet = map<identifier, symbolt>`).
+
+The abstract interface of `goto_modelt` is outlined in the file
+[`src/goto-programs/abstract_goto_model.h`](../../src/goto-programs/abstract_goto_model.h).
+
+## goto_functiont
+
+A `goto_functiont` is also a product type. It's designed to represent a function
+at the IR level, and effectively it's the following ADT (in pseudocode):
+
+```js
+type goto_functiont {
+    goto_programt body
+    list<identifiers> parameters
+}
+```
+
+Of the two types listed above, the `parameters` one should be self-documenting,
+(the values of these are later looked up in the symbol table), so we're going to
+focus next on the type `goto_programt`
+
+## goto_programt
+
+A `goto_programt` is a container of GOTO-IR instructions. In pseudocode, it would
+look like `type goto_programt = list<goto_instructiont>`.
+
+An instruction (`goto_instructiont`) is another product type, describing a GOTO level
+instruction with the following 3 component subtypes, again in pseudocode:
+
+```js
+type goto_instructiont {
+    instruction_type instr_type
+    instruction      statement
+    guard            boolean_expr
+}
+```
+
+The above subtypes are just illustrative, so we will provide some extra explanation for
+these:
+
+* The `instruction_type` above corresponds to the `goto_program_instruction_typet` type
+  listed in [`goto_program.h`](../../src/goto-programs/goto_program.h)
+  * For illustration purposes, some instruction types are `assign`, `function_call`, `return`,
+    etc.
+* The `instruction` is a statement represented by a `goto_instruction_codet`.
+  * A `goto_instruction_codet` is a `codet` (a data structure broadly representing a statement
+    inside CBMC) that contains the actual GOTO-IR instruction.
+  * You can distinguish different statements by inspecting the `irep` element `ID_statement`.
+* The `guard` is an `exprt` (A CBMC data structure broadly representing an expression)
+  that is expected to have type `bool`.
+  * This is optional - not every instruction is expected to have a `guard` associated with it.
+
+## source_locationt
+
+Another important data structure that needs to be discussed at this point is
+`source_locationt`.
+
+This is an `irept`. `irep`s are the central data structure that model most entities inside
+CBMC and the assorted tools - effectively a node/map like data structure that forms a hierachical
+tree that ends up modeling graphs like ASTs, CFGs, etc. (This will be further discussed in
+a dedicated page).
```

Comment:
I am not sure we should commit TODO notes.

Reply:
This was not so much as a TODO, rather, it's a small digression to guide user expectation (that they should expect a dedicated topic on this that they need to look for into the documentation).

Do you think this should leave, or is there a better way to phrase this?

Reply:
Maybe put in a link when it is available?

Reply:
I agree on not adding a TODO/promise. Maybe give a really brief overview and link to the source code/doxygen?